crates/project/src/lsp_command.rs

```diff
@@ -3178,3 +3183,477 @@ impl LspCommand for LinkedEditingRange {
+            .is_some()
+    }
+
+    fn to_lsp(
+        &self,
+        path: &Path,
+        _: &Buffer,
+        language_server: &Arc<LanguageServer>,
+        _: &AppContext,
+    ) -> Result<lsp::DocumentDiagnosticParams> {
+        let identifier = match language_server.capabilities().diagnostic_provider {
+            Some(lsp::DiagnosticServerCapabilities::Options(options)) => options.identifier.clone(),
+            Some(lsp::DiagnosticServerCapabilities::RegistrationOptions(options)) => {
+                options.diagnostic_options.identifier.clone()
+            }
+            None => None,
+        };
+
+        Ok(lsp::DocumentDiagnosticParams {
+            text_document: lsp::TextDocumentIdentifier {
+                uri: lsp::Url::from_file_path(path).unwrap(),
+            },
+            identifier,
+            previous_result_id: None,
+            partial_result_params: Default::default(),
+            work_done_progress_params: Default::default(),
+        })
+    }
+
+    async fn response_from_lsp(
+        self,
+        message: lsp::DocumentDiagnosticReportResult,
+        lsp_store: Model<LspStore>,
+        buffer: Model<Buffer>,
+        server_id: LanguageServerId,
+        mut cx: AsyncAppContext,
+    ) -> Result<Self::Response> {
+        let uri = buffer.read_with(&cx, |buffer, cx| {
+            buffer
+                .file()
+                .and_then(|file| file.as_local())
+                .map(|file| lsp::Url::from_file_path(file.abs_path(cx).clone()).unwrap())
+        })?;
+
+        let Some(uri) = uri else {
+            return Ok(None);
+        };
+
+        match message {
+            lsp::DocumentDiagnosticReportResult::Report(report) => match report {
+                lsp::DocumentDiagnosticReport::Full(report) => {
+                    lsp_store.update(&mut cx, |store, cx| {
+                        for (uri, report) in report.related_documents.into_iter().flatten() {
+                            match report {
+                                lsp::DocumentDiagnosticReportKind::Full(report) => {
+                                    store.update_diagnostics(
+                                        server_id,
+                                        lsp::PublishDiagnosticsParams {
+                                            diagnostics: report.items.clone(),
+                                            uri,
+                                            version: None,
+                                        },
+                                        &[],
+                                        cx,
+                                    ).expect("Failed to update diagnostics for the related document");
+                                }
+                                lsp::DocumentDiagnosticReportKind::Unchanged(_) => (),
+                            }
+                        }
+                    }).expect("Failed to update diagnostics for related documents");
```

Comment:
I'm concerned by the amount of new `.unwrap()`, `.expect(..)` in the new code, esp. given that this function returns `Result`.
Let's not write the code that may panic here, as it's quite a core feature, queried frequently, and we're parsing things from a potentially malicious server.

Reply:
Yes, totally makes sense. Changed that. The only ugly part left there is getting the URI of the document. We need that URI for `update_diagnostics` method on the `lsp_store`. If you have any suggestions for that part, please let me know. Thanks!